### DISCUSSION TITLE
Programmatic upload for related files

### DISCUSSION THREAD

**Discussion by onecrayon** [strapi/strapi#7706] [Legacy Topics]
My company is migrating a large existing database to Strapi, and I am having trouble figuring out how to manage the migration of images (and a few other files, like PDFs) into the new Strapi database. I've setup a script that reads in the legacy data and uses `strapi.query()` to populate data, which works great. However, while the docs mention that you can associate media by ID when running a `create` or `update`, it doesn't mention how to get the media into the system in the first place.

Does anyone have any insight into this topic? We've got thousands of images, and it just isn't practical to migrate them by hand but I've been banging my head against the problem for several days without any luck. At this point, I've got a bunch of remote URLs where the old files are living, and I know which stuff in the database they should be associated with, but have no way to migrate them into the upload plugin so that I can subsequently associate them with content.

Thanks in advance!

**Comment by derrickmehaffy** [comment]
@onecrayon in a case like yours I would suggest some intermediary script that handles reading from the old application and relays that data to the new application.

You are correct that we don't currently have good documentation for the internal upload API and it can be quite difficult to describe, I would say the best source of information is looking at the source directly: https://github.com/strapi/strapi/tree/master/packages/strapi-plugin-upload

We will be spending quite a bit of time over the next few months to update and refactor our documentation to make it much easier to find the information you are looking for.

**Reply by onecrayon** [reply]
> in a case like yours I would suggest some intermediary script that handles reading from the old application and relays that data to the new application.

Yep, this is exactly what I’m doing. My script bootstraps the basic Strapi shell, reads in CSV files of the old data (don’t ask 🤦‍♂️), then inserts them into the Strapi database using `strapi.query`. The one hang up is getting the images into Strapi (I have the URLs in the old data). Is there a particular entry point I should be looking at in the upload plugin? I actually browsed that codebase a little before posting (along with the front-end component), but am having a difficult time figuring out where it hooks up to Strapi (since I presumably need to be able to get to it from a Strapi object rather than `require`-ing in arbitrary pieces).

**Reply by derrickmehaffy** [reply]
In this case, I meant using the script and passing the information over the REST methods instead of directly hooking into the strapi shell.

Out of curiosity, would it not be possible to handle the migration at the database level? (Is your old one still in production, or do you have new data in the new app?)

**Reply by onecrayon** [reply]
@derrickmehaffy Ah, I see. So have the script POST to whatever URL is being used by the upload plugin. That would certainly work, albeit a lot more slowly; is there a particular file I should look at that defines the entrypoint URL for the upload plugin? I was browsing the front-end components with an eye to try and figure out this stuff, but they were Greek to me (suppose I could also just pop open the Inspector and seeing what the browser passes along; hadn't gone that far because I was hoping to be able to do things programmatically so that I could potentially do something like stream data straight from the old URL to Strapi which should drastically improve performance as I'm ultimately going to be handling thousands of files).

The old data is living in a decrepit Microsoft SQL server that's not accessible, which is why it ended up in CSV files. The new strapi app is completely empty, and I am comfortable interacting with SQL directly. If I directly INSERT rows in the files table, and then extract the IDs and setup the relationships by passing them to `strapi.query().update` via the script, will that be sufficient to establish everything database-side that strapi needs to operate normally?

**Reply by derrickmehaffy** [reply]
Yup! All routes used within Strapi are in a routes.json (always inside the config folder, but there are many all over)

https://github.com/strapi/strapi/blob/master/packages/strapi-plugin-upload/config/routes.json

Each route has a handler which is the Controller file and the method inside that file.

**Reply by onecrayon** [reply]
Excellent, thank you. I will take a gander and check back if it works/fails.

**Comment by onecrayon** [comment]
Darn it, I completely forgot to update this issue. So the suggestion of posting files to a locally-hosted Strapi worked out alright. Here's a high-level snippet that gives a sketch of how I handled the problem (including some working Strapi code, if anyone else runs across this problem in the future):

```js
// First installed dependencies: `npm install form-data`
const fs = require('fs')
const bootstrapStrapi = require('strapi')
const FormData = require('form-data')

// This stuff needs to be in an async function to avoid Promise hell (pretty much anything
// database-related is async)
(async () => {
  // I passed in my bearer token via command line argument for this script
  const token = process.argv[2]
  // Setup our Strapi instance; environment variables are in `.env` and loaded automatically
  const strapi = bootstrapStrapi()
  await strapi.load()
  
  console.log('START linking files...')

  // I gathered a list of local files via fs module into this variable
  const files = ['file.jpg']
  for (const filename of files) {
    const filepath = `/path/to/${filename}`
    if (!fs.existsSync(filepath)) continue
    let fileId
    // Verify we haven't already uploaded this file
    const existing = await strapi.plugins.upload.models.file.where('name', '=', filename).fetch({
      require: false,
      columns: 'id',
      withRelated: ''
    })
    if (existing) {
      fileId = existing.id
    } else {
      // Upload the file
      const form = new FormData()
      form.append('files', fs.createReadStream(filepath))
      form.append('fileInfo', JSON.stringify({
        "alternativeText": "",
        "caption": "",
        "name": null
      }))
      const response = await (new Promise((resolve, reject) => {
        form.submit({
          'host': 'localhost',
          'port': 1337,
          'path': '/upload',
          'headers': {
            'Authorization': `Bearer ${token}`
          }
        }, (err, res) => {
          if (err) reject(err)
          else resolve(res)
        })
      })).catch(err => {
        console.log(`Failed to upload ${filename}: ${err}`)
      })
      if (!response) continue
      // We need to get the JSON response back, so wrap the callback definitions in a promise
      let data = await (new Promise((resolve, reject) => {
        let body = ''
        response.on('data', (chunk) => {
          body += chunk;
        })
        response.on('end', () => {
          if (response.statusCode !== 200) reject(`${response.statusCode} - ${body}`)
          else resolve(body)
        })
        response.on('error', err => {
          reject(err)
        })
      })).catch(err => {
        console.log(`Failed to get JSON response ${filename}: ${err}`)
      })
      if (!data) continue
      data = JSON.parse(data)
      fileId = data[0].id
    }
    if (fileId === undefined) continue
    // At this point, we've got the file ID, and can attach that to existing models.
    // How you go about this depends a lot on your individual Strapi data setup, but
    // generally you'll probably do something like:
    await strapi.query('my-model').update({ 'id': 'my-model-ID' }, {'image': fileId}}, [])
  }
})().catch((e) => { console.error(e) }).finally(() => {
  console.log('FINISHED linking files.')
  process.exit()
})
```
